### PR TITLE
Add feature to download mp4 videos or gifs from subreddit

### PR DESCRIPTION
**Pull request by zzLinus**

### CODE REVIEW

**Comment by zzLinus** [comment]
Hi, I'm pretty new to open source project contribution....So if any thing i did know please let me know

File: extractors/reddit/reddit.go [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+		for i := len(url) - 1; i >= 0; i-- {
+			if url[i] == ';' {
+				urlL = url[i+1 : len(url)-1]
+				break
+			}
+		}
+
+		url = urlU + urlL
+
+		url = fmt.Sprintf("%s", url)
```
Comment (iawia002) [inline extractors/reddit/reddit.go:137]: What's the original URL might be, are you trying to remove some characters from the URL?

File: extractors/reddit/reddit_test.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+		args     test.Args
+		playlist bool
+	}{
+		{
+			name: "normal test 1",
+			args: test.Args{
+				URL:   "https://www.reddit.com/r/space/comments/uj8sod/a_couple_of_days_ago_i_visited_this_place_an/",
+				Title: "A couple of days ago I visited this place. An abandoned space shuttle : space",
+			},
+			playlist: false,
```
Comment (iawia002) [inline extractors/reddit/reddit_test.go:24]:
```suggestion
```

File: extractors/reddit/reddit_test.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+		},
+		{
+			name: "normal test 2",
+			args: test.Args{
+				URL:   "https://www.reddit.com/r/DotA2/comments/uq012r/til_how_useful_hurricane_bird_is/",
+				Title: "TIL how useful hurricane bird is : DotA2",
+			},
+			playlist: false,
```
Comment (iawia002) [inline extractors/reddit/reddit_test.go:32]:
```suggestion
```

File: extractors/reddit/reddit_test.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+		t.Run(tt.name, func(t *testing.T) {
+			var (
+				data []*extractors.Data
+				err  error
+			)
+			data, err = New().Extract(tt.args.URL, extractors.Options{})
+			fmt.Print(data)
+			if err != nil {
+				log.Fatal(err)
+			}
```
Comment (iawia002) [inline extractors/reddit/reddit_test.go:45]: Do not print the data and do `log.Fatal` here, eg: https://github.com/iawia002/lux/blob/b3e156d2e1113d7f2c047efae7561b1beceaeeae/extractors/douyin/douyin_test.go#L31-L35

**Comment by iawia002** [comment]
And you should update the README: https://github.com/iawia002/lux#supported-sites.

One more question: do we support downloading typical images, not just gifs?

**Comment by zzLinus** [comment]
Thanks for the instructions！I didn't add image support yet,because Reddit mobile app can download image very easily.But it should be easy to add this feature,just give me one day or less i think it should be enough◑.◑

File: extractors/reddit/reddit.go [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+				Title:   videoName,
+				Type:    extractors.DataTypeAudio,
+				Streams: streams,
+				URL:     "www.reddit.com",
+			},
+		}
+	} else if fileType == "gif" {
+		var (
+			urlU string
+			urlL string
```
[the remark this answers is not in the document]
Comment (zzLinus) [reply extractors/reddit/reddit.go:116]: The url string match by regular expresion:
// https://preview.redd.it/t939jge21sz81.gif?format=mp4&amp;s=8c72871c30d5a5e3f265e2c1363b45b63bf044cd"
The url which needed for download the real content:
// https://preview.redd.it/t939jge21sz81.gif?format=mp4&s=8c72871c30d5a5e3f265e2c1363b45b63bf044cd
so the urlU(url up) -> https://preview.redd.it/t939jge21sz81.gif?format=mp4&
urlL(url low) -> s=8c72871c30d5a5e3f265e2c1363b45b63bf044cd
I know it's kind of stupid but the json datas provided by reddit use  a postID as key i don't know how to unmarshal it so this part  is really mess up...I mean i can extract the postID but i don't know how to use it parse the json....

[the remark this answers is not in the document]
Comment (iawia002) [reply extractors/reddit/reddit.go:116]: You can simply replace the special characters in the URL:

```go
strings.ReplaceAll("https://preview.redd.it/t939jge21sz81.gif?format=mp4&amp;s=8c72871c30d5a5e3f265e2c1363b45b63bf044cd", "&amp;", "&")
```

File: README.md [hunk 6]
```diff
@@ -628,6 +628,7 @@ $ lux -j "https://www.bilibili.com/video/av20203945"
 | 虎牙视频   | <https://v.huya.com>         | ✓         |           |          |             |                   | [![huya](https://github.com/iawia002/lux/actions/workflows/stream_huya.yml/badge.svg)](https://github.com/iawia002/lux/actions/workflows/stream_huya.yml)                   |
 | 喜马拉雅   | <https://www.ximalaya.com>   |           |           | ✓        |             |                   | [![ximalaya](https://github.com/iawia002/lux/actions/workflows/stream_ximalaya.yml/badge.svg)](https://github.com/iawia002/lux/actions/workflows/stream_ximalaya.yml)       |
 | 快手       | <https://www.kuaishou.com>   | ✓         |           |          |             |                   | [![kuaishou](https://github.com/iawia002/lux/actions/workflows/stream_kuaishou.yml/badge.svg)](https://github.com/iawia002/lux/actions/workflows/stream_kuaishou.yml)       |
+| Reddit     | <https://www.reddit.com>     | ✓         |           | ✓        |             |                   | [![reddit](https://github.com/iawia002/lux/actions/workflows/stream_reddit.yml/badge.svg)](https://github.com/iawia002/lux/actions/workflows/stream_reddit.yml)       |
```
Comment (iawia002) [inline README.md:631]:
```suggestion
| Reddit     | <https://www.reddit.com>     | ✓         |  ✓        |          |             |                   | [![reddit](https://github.com/iawia002/lux/actions/workflows/stream_reddit.yml/badge.svg)](https://github.com/iawia002/lux/actions/workflows/stream_reddit.yml)       |
```

The `Image` field should be checked, not `Audio`

File: extractors/reddit/reddit.go [hunk 7]
```diff
@@ -11,7 +11,6 @@ import (
 )
 
 var (
-	PostId       string
 	redditMP4API = "https://v.redd.it/"
 	audioURLPart = "/DASH_audio.mp4"
```
Comment (zzLinus) [inline extractors/reddit/reddit.go:15]: I read the discussion here https://github.com/iawia002/lux/issues/38 and i didn't figure out how to implemented it yet

File: extractors/reddit/reddit.go [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+func init() {
+	extractors.Register("reddit", New())
+}
+
+func New() extractors.Extractor {
+	return &extractor{}
+}
+
+func (e *extractor) Extract(url string, option extractors.Options) ([]*extractors.Data, error) {
+	var err error
```
Comment (iawia002) [inline extractors/reddit/reddit.go:35]:
```suggestion
```

File: extractors/reddit/reddit.go [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+		return nil, errors.WithStack(err)
+	}
+
+	// set thread number to 1 manually to avoid http 412 error
+	option.ThreadNumber = 1
+
+	return extractContent(url, html, option)
+}
+
+func extractContent(url, html string, extractOption extractors.Options) ([]*extractors.Data, error) {
```
Comment (iawia002) [inline extractors/reddit/reddit.go:47]: Seems no need to define a separate function

File: extractors/reddit/reddit.go [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+				URL:     url,
+			},
+		}, nil
+	} else if fileType == "gif" {
+		gifURL := utils.MatchOneOf(html, `https:\/\/preview\.redd\.it\/.*?\.gif\?format=mp4.*?"`)[0]
+		if gifURL == "" {
+			return nil, errors.New("can't match gif content downloadable url")
+		}
+
+		strings.ReplaceAll(gifURL, "&amp;", "&")
```
Comment (iawia002) [inline extractors/reddit/reddit.go:110]:
```suggestion
		gifURL = strings.ReplaceAll(gifURL, "&amp;", "&")
```

File: extractors/reddit/reddit.go [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+		fileType = "gif"
+	}
+
+	if fileType == "mp4" {
+		url := utils.MatchOneOf(html, `https://v.redd.it/(.+?)/HLSPlaylist`)[1]
+		if url == "" {
+			panic("can't match anything")
+		}
+
+		videoURL := fmt.Sprintf("%s%s%s", redditMP4API, url, res720)
```
[the remark this answers is not in the document]
Comment (zzLinus) [reply extractors/reddit/reddit.go]: I made a new commit to support different resolution

File: extractors/reddit/reddit.go [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+	}
+
+	if fileType == "mp4" {
+		mp4URL := utils.MatchOneOf(html, `https://v.redd.it/(.+?)/HLSPlaylist`)[1]
+		if mp4URL == "" {
+			return nil, errors.New("can't match mp4 content downloadable url")
+		}
+
+		streams := make(map[string]*extractors.Stream, len(resURLParts))
+		for id := 0; id < len(resURLParts); id++ {
```
Comment (iawia002) [inline extractors/reddit/reddit.go:60]: Why not turn `resURLParts` and `res` to a map:

```go
resURLParts := map[string]string{
    "720p": "/DASH_720.mp4",
}
```

File: extractors/reddit/reddit.go [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+			index := strconv.Itoa(id)
+			id, err := strconv.Atoi(index)
+			if err != nil {
+				return nil, err
+			}
+			resURL := fmt.Sprintf("%s%s%s", redditMP4API, mp4URL, resURLParts[id])
+			audioURL := fmt.Sprintf("%s%s%s", redditMP4API, mp4URL, audioURLPart)
```
Comment (iawia002) [inline extractors/reddit/reddit.go:67]: If all videos share one audio file, then you can generate a separate audio part without having to do it inside the loop

File: extractors/reddit/reddit.go [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+				Parts: contentData,
+				Size:  videoSize,
+			},
+		}
+
+		return []*extractors.Data{
+			{
+				Site:    "Reddit reddit.com",
+				Title:   contentName,
+				Type:    extractors.DataTypeVideo,
```
Comment (iawia002) [inline extractors/reddit/reddit.go:139]: The type should be `DataTypeImage`

File: extractors/reddit/reddit.go [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+				Streams: streams,
+				URL:     url,
+			},
+		}, nil
+	} else if fileType == "img" {
+		var imgURL string
+		var is int64
```
Comment (iawia002) [inline extractors/reddit/reddit.go:146]: What does `is` mean?

File: extractors/reddit/reddit.go [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+			URL:  imgURL,
+			Size: is,
+			Ext:  "jpg",
+		})
+
+		return []*extractors.Data{
+			{
+				Site:  "Reddit reddit.com",
+				Title: contentName,
+				Type:  extractors.DataTypeVideo,
```
Comment (iawia002) [inline extractors/reddit/reddit.go:175]:
```suggestion
				Type:  extractors.DataTypeImage,
```

File: extractors/reddit/reddit.go [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+
+		videoSize, err := request.Size(gifURL, "reddit.com")
+		if err != nil {
+			return nil, errors.New("can't get video size")
+		}
+		contentData := make([]*extractors.Part, 0)
+		contentData = append(contentData, &extractors.Part{
+			URL:  gifURL,
+			Size: videoSize,
+			Ext:  "mp4",
```
Comment (iawia002) [inline extractors/reddit/reddit.go:125]: Why is the ext is mp4, shouldn't it be a gif?

File: extractors/reddit/reddit.go [hunk 18]
```diff
@@ -0,0 +1,187 @@
...
+		})
+
+		streams := map[string]*extractors.Stream{
+			"video": {
```
Comment (iawia002) [inline extractors/reddit/reddit.go:129]: The key means quality not type, if there is only one quality, I prefer `default`

File: extractors/reddit/reddit.go [hunk 19]
```diff
@@ -0,0 +1,187 @@
...
+			Ext:  "jpg",
+		})
+
+		return []*extractors.Data{
+			{
+				Site:  "Reddit reddit.com",
+				Title: contentName,
+				Type:  extractors.DataTypeVideo,
+				Streams: map[string]*extractors.Stream{
+					"image": {
```
Comment (iawia002) [inline extractors/reddit/reddit.go:177]: ditto

File: extractors/reddit/reddit_test.go [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			var (
+				data []*extractors.Data
+				err  error
+			)
+			data, err = New().Extract(tt.args.URL, extractors.Options{})
```
Comment (iawia002) [inline extractors/reddit/reddit_test.go:36]:
```suggestion
			data, err := New().Extract(tt.args.URL, extractors.Options{})
```

File: extractors/reddit/reddit.go (hunk 15 above)
Comment (zzLinus) [reply extractors/reddit/reddit.go:146]: it's imagesize......really hard to name variable sometimes...

File: extractors/reddit/reddit.go (hunk 14 above)
Comment (zzLinus) [reply extractors/reddit/reddit.go:139]: the gifs need to be typed as mp4 in order to play,if i name this gifs file as something like demo.gif it won't play...So do i type is as DataTypeVideos...

File: extractors/reddit/reddit.go [hunk 21]
```diff
@@ -0,0 +1,187 @@
...
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/iawia002/lux/extractors"
+	"github.com/iawia002/lux/request"
+	"github.com/iawia002/lux/utils"
+	"github.com/pkg/errors"
```
[the remark this answers is not in the document]
Comment (zzLinus) [reply extractors/reddit/reddit.go:11]: Does that mean 
```
"github.com/pkg/errors"

"github.com/iawia002/lux/extractors"
"github.com/iawia002/lux/request"
"github.com/iawia002/lux/utils"
```

[the remark this answers is not in the document]
Comment (iawia002) [reply extractors/reddit/reddit.go:11]: Yes, I mentioned here https://github.com/iawia002/lux/pull/1103#discussion_r885176096

[the remark this answers is not in the document]
Comment (zzLinus) [reply extractors/reddit/reddit.go:11]: Oh sorry,i forgot that.Will go compiler compile faster if packages has been grouped?

File: extractors/reddit/reddit.go (hunk 17 above)
Comment (zzLinus) [reply extractors/reddit/reddit.go:125]: The gif downloadable link provide by reddit is a mp4 file,it's kind of weird

File: extractors/reddit/reddit.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (iawia002) [reply extractors/reddit/reddit.go]: No, it's just that we need to follow the same code style.

File: extractors/reddit/reddit.go [hunk 22]
```diff
@@ -0,0 +1,190 @@
...
+	}
+
+	if fileType == "mp4" {
+		mp4URL := utils.MatchOneOf(html, `https://v.redd.it/(.+?)/HLSPlaylist`)[1]
+		if mp4URL == "" {
+			return nil, errors.New("can't match mp4 content downloadable url")
+		}
+
+		audioURL := fmt.Sprintf("%s%s%s", redditMP4API, mp4URL, audioURLPart)
+		size, err := request.Size(audioURL, referer)
```
Comment (iawia002) [inline extractors/reddit/reddit.go:58]:
```go
if err != nil {
	return nil, err
}
```

File: extractors/reddit/reddit.go [hunk 23]
```diff
@@ -0,0 +1,190 @@
...
+		audioPart := &extractors.Part{
+			URL:  audioURL,
+			Size: size,
+			Ext:  "mp3",
+		}
+
+		resMap := map[string]string{
```
Comment (iawia002) [inline extractors/reddit/reddit.go:65]: This can be defined outside this function

File: extractors/reddit/reddit.go [hunk 24]
```diff
@@ -0,0 +1,190 @@
...
+			"480p": "/DASH_480.mp4",
+			"360p": "/DASH_360.mp4",
+			"240p": "/DASH_240.mp4",
+			"220p": "/DASH_220.mp4",
+		}
+		streams := make(map[string]*extractors.Stream, len(resMap))
+		for res, urlparts := range resMap {
+			if err != nil {
+				return nil, err
+			}
```
Comment (iawia002) [inline extractors/reddit/reddit.go:76]:
```suggestion
```

File: extractors/reddit/reddit.go [hunk 25]
```diff
@@ -0,0 +1,190 @@
...
+			resURL := fmt.Sprintf("%s%s%s", redditMP4API, mp4URL, urlparts)
+			size, err := request.Size(resURL, referer)
+			if err != nil {
+				return nil, err
+			}
+			if err != nil {
+				return nil, err
+			}
```
Comment (iawia002) [inline extractors/reddit/reddit.go:84]:
```suggestion
			if err != nil {
				return nil, err
			}
```

File: extractors/reddit/reddit.go [hunk 26]
```diff
@@ -0,0 +1,190 @@
...
+			Size: videoSize,
+			Ext:  "gif",
+		})
+
+		streams := map[string]*extractors.Stream{
+			"default": {
+				Parts: contentData,
+				Size:  videoSize,
+			},
+		}
```
Comment (iawia002) [inline extractors/reddit/reddit.go:138]: You don't have to define a `contentData` variable

```suggestion
		streams := map[string]*extractors.Stream{
			"default": {
				Parts: []*extractors.Part{{...}},
				Size:  videoSize,
			},
		}
```

File: extractors/reddit/reddit.go [hunk 27]
```diff
@@ -0,0 +1,190 @@
...
+		return []*extractors.Data{
+			{
+				Site:  "Reddit reddit.com",
+				Title: contentName,
+				Type:  extractors.DataTypeImage,
+				Streams: map[string]*extractors.Stream{
+					"default": {
+						Parts: imgParts,
+						Size:  size,
+					},
```
Comment (iawia002) [inline extractors/reddit/reddit.go:183]: ditto

**Review by iawia002 — APPROVED** [review]
LGTM, thanks!

**Pull request merged**